File: .before-java-project/src/test/java/com/github/te170476/musicdsl/score/note/NoteValueTest.java
package com.github.te170476.musicdsl.score.note;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class NoteValueTest {
    INoteValue whole = NoteValue.get(1);
    INoteValue triplet = NoteValue.get(3).setBase(whole);
    INoteValue nonuplet = NoteValue.get(9).setBase(whole);

    @Test
    void relativeEqualsAbsolute() {
        var relativeNonuplet = NoteValue.get(3).setBase(triplet);
        var absoluteNonuplet = nonuplet;
        assertEquals(
                relativeNonuplet.toPercentage(),
                absoluteNonuplet.toPercentage()
        );

        var relativeTriplet = NoteValue.get(nonuplet, nonuplet, nonuplet);
        var absoluteTriplet = triplet;
        assertEquals(
                relativeTriplet.toPercentage(),
                absoluteTriplet.toPercentage()
        );
    }
}

File: src/test/scala/com/github/nnnnusui/musicdsl/entity/score/note/NoteValueTest.scala
package com.github.nnnnusui.musicdsl.entity.score.note

import org.scalatest.matchers.should.Matchers
import org.scalatest.wordspec.AnyWordSpec

class NoteValueTest extends AnyWordSpec with Matchers {
  private val whole = NoteValue(1)
  private val triplet = NoteValue(3) / whole
  private val nonuplet = NoteValue(9) / whole

  "NoteValue" should {
    "3.base(3) equals 9" in {
      (triplet / triplet).percentage should ===(nonuplet.percentage)
    }
    "(9, 9, 9) equals 3" in {
      (nonuplet * 3).percentage should ===(triplet.percentage)
    }
  }
}
